### PR TITLE
Disallow SpreadField in table constructor

### LINKED ISSUE #36474: Cannot instantiate table with literal
**Description:**

Trying to instantiate a table with literal causes an error.

**Steps to reproduce:**
~~~ballerina
Book volleyballHandbook = {
    isbn: "978-0736056106",
    title: "The Volleyball Handbook",
    author_id: "bob-miller"
};

Book friendshipBread = {
    isbn: "978-0345525345",
    title: "Friendship Bread",
    author_id: "darien-gee"
};

table<Book> key(isbn) bookTable1 = table [{...volleyballHandbook}, {...friendshipBread}];
~~~

Here is the stacktrace:

~~~
java.lang.NullPointerException
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.checkForDuplicateKeys(DataflowAnalyzer.java:1007)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.visit(DataflowAnalyzer.java:994)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangTableConstructorExpr.accept(BLangTableConstructorExpr.java:70)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.analyzeNode(DataflowAnalyzer.java:2268)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.visit(DataflowAnalyzer.java:666)
        at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:54)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.analyzeNode(DataflowAnalyzer.java:2268)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.visit(DataflowAnalyzer.java:323)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.analyzeNode(DataflowAnalyzer.java:2268)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.DataflowAnalyzer.analyze(DataflowAnalyzer.java:293)
        at io.ballerina.projects.internal.CompilerPhaseRunner.dataflowAnalyze(CompilerPhaseRunner.java:200)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:124)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:426)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:376)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:200)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:185)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:101)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:96)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:214)
        at io.ballerina.projects.Package.getCompilation(Package.java:141)
        at io.ballerina.shell.invoker.ShellSnippetsInvoker.compile(ShellSnippetsInvoker.java:235)
        at io.ballerina.shell.invoker.classload.ClassLoadInvoker.getCompilation(ClassLoadInvoker.java:242)
        at io.ballerina.shell.EvaluatorImpl.getCompilation(EvaluatorImpl.java:89)
        at io.ballerina.shell.cli.BallerinaShell.run(BallerinaShell.java:121)
        at io.ballerina.shell.cli.ReplShellApplication.execute(ReplShellApplication.java:84)
        at io.ballerina.cli.cmd.ShellCommand.execute(ShellCommand.java:79)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
| Something went wrong: java.lang.IllegalArgumentException: count is negative: -4
~~~
**Affected Versions:**
Ballerina 2201.0.4 (Swan Lake)
Language specification 2022R1
**OS, DB, other environment details and versions:**
MacOS
**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes #36474 

```
type FooRec record {
    readonly int x;
    int y;
};

FooRec spreadField1 = {x: 1002, y: 30};
FooRec spreadField2 = {x: 1003, y: 25};

table<FooRec> key(x) tb1 = table [
            {x: 1001, y: 20},
            {...spreadField1},
            {...spreadField2}
        ];
```
This should give `value expression of key specifier 'x' must be a constant expression` as error.
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java [hunk 1]
```diff
@@ -155,7 +155,11 @@ public void testTableNegativeCases() {
         validateError(compileResult, index++, "incompatible types: expected 'int', found '[int,string,string]'",
                 462, 21);
         validateError(compileResult, index++, "incompatible types: expected 'int', found '[int,int,int,int]'", 469, 21);
-        validateError(compileResult, index, "incompatible types: expected 'int', found '[int,int,int]'", 478, 21);
+        validateError(compileResult, index++, "incompatible types: expected 'int', found '[int,int,int]'", 478, 21);
+        validateError(compileResult, index++, "value expression of key specifier 'x' must be a constant expression",
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java:159]: We should get this error for both L:491 and L:492 right?

Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/table/TableNegativeTest.java:159]: We should get this error for both L:491 and L:492 right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 2]
```diff
@@ -1354,6 +1354,8 @@ private boolean validateTableKeyValue(List<String> keySpecifierFieldNames,
                 dlog.error(recordLiteral.pos,
                         DiagnosticErrorCode.KEY_SPECIFIER_FIELD_VALUE_MUST_BE_CONSTANT_EXPR, fieldName);
                 data.resultType = symTable.semanticError;
+            }
+            if (data.resultType == symTable.semanticError) {
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:1358]: It seems this is not a correct thing to do,
Consider this example.
```ballerina
table<FooRec> key(x, y) tb1 = table [
            {x: 1001, y: 20},
            {x: i, y: i}
        ];
```
(Note that there are two keys)
For this we are getting only one error, but we should get two errors for two keys.

I think the correct thing should be,
```java
    private boolean validateTableKeyValue(List<String> keySpecifierFieldNames,
                                                           List<BLangRecordLiteral> recordLiterals, AnalyzerData data) {

        for (String fieldName : keySpecifierFieldNames) {
            for (BLangRecordLiteral recordLiteral : recordLiterals) {
                BLangExpression recordKeyValueField = getRecordKeyValueField(recordLiteral, fieldName);
                if (recordKeyValueField != null && isConstExpression(recordKeyValueField)) {
                    continue;
                }

                dlog.error(recordLiteral.pos,
                        DiagnosticErrorCode.KEY_SPECIFIER_FIELD_VALUE_MUST_BE_CONSTANT_EXPR, fieldName);
                data.resultType = symTable.semanticError;
            }
        }

        return data.resultType != symTable.semanticError;
    }
```

With this change, if we have two key and spreadField, we are getting 4 errors(2 for one key). I think that is fine.
@lasinicl , @pcnfernando WDYT?

**Review by KavinduZoysa — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/types/table/table-negative.bal [hunk 3]
```diff
@@ -477,3 +477,35 @@ function testMultipleKeys3() {
...
+table<FooRec> key(x) tb1 = table [
+            {x: 1001, y: 20},
+            {...spreadField1},
+            {...spreadField2}
+        ];
+
+table<FooRec> key(x) tb2 = table [
+            {...spreadField1},
+            {...spreadField2}
+        ];
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/table/table-negative.bal:498]: Shall We follow same indentation like `FooRec` in L481.
ex:- 
```
table<FooRec> key(x) tb2 = table [
    {...spreadField1},
    {...spreadField2}
];
```

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/table/table-negative.bal:498]: Yes. @prakanth97 Intellij Idea will not format Ballerina code properly at the moment, Please use VS Code to format the code

**Review by KavinduZoysa — APPROVED** [review]

**Review by gimantha — APPROVED** [review]

**Comment by gimantha** [comment]
Shall we add a test case for spread operator where we don't have key specifier?

**Comment by gimantha** [comment]
@prakanth97 Shall we send the same PR to 2201.1.x and 2201.0.x branches as well?

**Pull request merged**